File: boilerplate/python/order-service/order-service-python/src/application/usecases/place_order_use_case_impl.py
from .place_order_use_case import PlaceOrderUseCase
from ..domain.services.order_placement_service import OrderPlacementService
from ..domain.ports.order_repository import OrderRepository
from .dtos.order_dtos import CreateOrderCommand, OrderResult
from loguru import logger
from libs.python-common.src.common.infrastructure.api.middleware import log_use_case

class PlaceOrderUseCaseImpl(PlaceOrderUseCase):
    def __init__(self, order_placement_service: OrderPlacementService):
        self.order_placement_service = order_placement_service

    @log_use_case
    async def execute(self, command: CreateOrderCommand) -> OrderResult:
        # Map DTO to Domain
        from .dtos.order_dtos import OrderItemDTO
        from ..domain.models.order import OrderItem

        items = [
            OrderItem(
                product_id=item.product_id,
                quantity=item.quantity,
                unit_price=item.unit_price
            ) for item in command.items
        ]

        # Delegate to Domain Service for business logic
        order = self.order_placement_service.place_order(command.customer_id, items)

        return OrderResult(
            order_id=order.id.value,
            status=order.status,
            created_at=order.created_at.isoformat()
        )

File: boilerplate/java/order-service/src/main/java/com/example/orderservice/application/usecases/PlaceOrderUseCaseImpl.java
package com.example.orderservice.application.usecases;

import com.example.orderservice.application.dtos.*;
import com.example.orderservice.domain.events.OrderPlaced;
import com.example.orderservice.domain.models.*;
import com.example.orderservice.domain.ports.EventPublisher;
import com.example.orderservice.domain.services.OrderPlacementService;

import java.math.BigDecimal;
import java.util.List;
import java.util.stream.Collectors;

public class PlaceOrderUseCaseImpl implements PlaceOrderUseCase {

    private final OrderPlacementService orderPlacementService;
    private final EventPublisher eventPublisher;

    public PlaceOrderUseCaseImpl(
        OrderPlacementService orderPlacementService,
        EventPublisher eventPublisher
    ) {
        this.orderPlacementService = orderPlacementService;
        this.eventPublisher = eventPublisher;
    }

    @Override
    public OrderResult execute(CreateOrderCommand command) {
        if (command == null) {
            throw new IllegalArgumentException("command must not be null");
        }
        List<OrderItem> items = command.items().stream()
            .map(dto -> new OrderItem(dto.productId(), dto.quantity(), dto.unitPrice()))
            .collect(Collectors.toList());
        
        Order order = orderPlacementService.placeOrder(command.customerId(), items);
        
        // Publish domain event
        OrderPlaced.OrderItemEventData[] itemData = order.getItems().stream()
            .map(item -> new OrderPlaced.OrderItemEventData(
                item.getProductId(),
                item.getQuantity(),
                item.getUnitPrice(),
                item.getTotalAmount()
            ))
            .toArray(OrderPlaced.OrderItemEventData[]::new);
        
        OrderPlaced event = new OrderPlaced(
            order.getId(),
            order.getCustomerId(),
            List.of(itemData),
            order.getTotalAmount()
        );
        eventPublisher.publish(event);
        
        return new OrderResult(order.getId().getValue(), order.getStatus().name(), order.getCreatedAt());
    }
}
